// enqueuer/js/dynamic-functions/dynamic-function-controller.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const logger_1 = require("../loggers/logger");
class DynamicFunctionController {
    constructor(functionBody, thisArg = null) {
        this.arguments = [];
        this.functionBody = functionBody;
        this.thisArg = thisArg;
        this.addArgument('require', require);
        this.addArgument('Logger', logger_1.Logger);
    }
    addArgument(name, value) {
        this.arguments.push({ name: name, value: value });
    }
    execute() {
        const createdFunction = this.createFunction();
        return this.executeFunction(createdFunction);
    }
    createFunction() {
        try {
            const constructorArgs = this.arguments.map(arg => arg.name).concat(this.functionBody);
            return ((...args) => new Function(...args)).apply(this.thisArg, constructorArgs);
        }
        catch (err) {
            logger_1.Logger.error(`Error creating function '${err}'`);
            throw err;
        }
    }
    executeFunction(dynamicFunction) {
        try {
            const callArgs = this.arguments.map(arg => arg.value);
            return dynamicFunction.apply(this.thisArg, callArgs);
        }
        catch (err) {
            logger_1.Logger.error(`Error running function '${err}'}`);
            throw err;
        }
    }
}
exports.DynamicFunctionController = DynamicFunctionController;
