// enqueuer/js/asserters/expect-to-be-greater-than-asserter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class ExpectToBeGreaterThanAsserter {
    assert(assertion, literal) {
        const name = assertion.name;
        const actual = assertion.expect;
        const expected = assertion.toBeGreaterThan;
        return {
            name,
            valid: assertion.not === undefined ? actual > expected : actual <= expected,
            description: `Expected '${literal.expect}'${assertion.not === undefined ?
                ' not' : ''} to be greater than '${expected}'. Received '${actual}'`
        };
    }
}
exports.ExpectToBeGreaterThanAsserter = ExpectToBeGreaterThanAsserter;
function entryPoint(mainInstance) {
    mainInstance.asserterManager.addAsserter({
        expect: {
            type: 'number',
            description: 'actual value'
        }, not: {
            required: false,
            description: 'negates',
            type: 'null'
        }, toBeGreaterThan: {
            type: 'number',
            description: 'expected value'
        }
    }, () => new ExpectToBeGreaterThanAsserter());
}
exports.entryPoint = entryPoint;
