// enqueuer/js/asserters/expect-to-be-less-than-asserter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class ExpectToBeLessThanAsserter {
    assert(assertion, literal) {
        const name = assertion.name;
        const actual = assertion.expect;
        const expected = assertion.toBeLessThan;
        return {
            name,
            valid: assertion.not === undefined ? actual < expected : actual >= expected,
            description: `Expected '${literal.expect}'${assertion.not !== undefined ?
                ' not' : ''} to be less than '${expected}'. Received '${actual}'`
        };
    }
}
exports.ExpectToBeLessThanAsserter = ExpectToBeLessThanAsserter;
function entryPoint(mainInstance) {
    mainInstance.asserterManager.addAsserter({
        expect: {
            type: 'number',
            description: 'actual value'
        }, not: {
            required: false,
            description: 'negates',
            type: 'null'
        }, toBeLessThan: {
            type: 'number',
            description: 'expected value'
        }
    }, () => new ExpectToBeLessThanAsserter());
}
exports.entryPoint = entryPoint;
