// enqueuer/js/asserters/expect-to-be-less-than-or-equal-to-asserter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class ExpectToBeLessThanOrEqualToAsserter {
    assert(assertion, literal) {
        const name = assertion.name;
        const actual = assertion.expect;
        const expected = assertion.toBeLessThanOrEqualTo;
        return {
            name,
            valid: assertion.not === undefined ? actual <= expected : actual > expected,
            description: `Expected '${literal.expect}'${assertion.not === undefined ?
                ' not' : ''} to be less than or equal to '${expected}'. Received '${actual}'`
        };
    }
}
exports.ExpectToBeLessThanOrEqualToAsserter = ExpectToBeLessThanOrEqualToAsserter;
function entryPoint(mainInstance) {
    mainInstance.asserterManager.addAsserter({
        expect: {
            type: 'number',
            description: 'actual value'
        }, not: {
            required: false,
            description: 'negates',
            type: 'null'
        }, toBeLessThanOrEqualTo: {
            type: 'number',
            description: 'expected value'
        }
    }, () => new ExpectToBeLessThanOrEqualToAsserter());
}
exports.entryPoint = entryPoint;
