// enqueuer/js/asserters/expect-to-be-greater-than-or-equal-to-asserter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class ExpectToBeGreaterThanOrEqualToAsserter {
    assert(assertion, literal) {
        const name = assertion.name;
        const actual = assertion.expect;
        const expected = assertion.toBeGreaterThanOrEqualTo;
        return {
            name,
            valid: assertion.not === undefined ? actual >= expected : actual < expected,
            description: `Expected '${literal.expect}'${assertion.not !== undefined ?
                ' not' : ''} to be greater than or equal to '${expected}'. Received '${actual}'`
        };
    }
}
exports.ExpectToBeGreaterThanOrEqualToAsserter = ExpectToBeGreaterThanOrEqualToAsserter;
function entryPoint(mainInstance) {
    mainInstance.asserterManager.addAsserter({
        expect: {
            type: 'number',
            description: 'actual value'
        }, not: {
            required: false,
            description: 'negates',
            type: 'null'
        }, toBeGreaterThanOrEqualTo: {
            type: 'number',
            description: 'expected value'
        }
    }, () => new ExpectToBeGreaterThanOrEqualToAsserter());
}
exports.entryPoint = entryPoint;
