// enqueuer/js/publishers/standard-output-publisher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const publisher_1 = require("./publisher");
const publisher_protocol_1 = require("../protocols/publisher-protocol");
class StandardOutputPublisher extends publisher_1.Publisher {
    constructor(publisherProperties) {
        super(publisherProperties);
    }
    publish() {
        if (typeof (this.payload) === 'object') {
            this.payload = JSON.stringify(this.payload, null, 2);
        }
        console.log(this.payload);
        return Promise.resolve();
    }
}
function entryPoint(mainInstance) {
    const protocol = new publisher_protocol_1.PublisherProtocol('stdout', (publisherModel) => new StandardOutputPublisher(publisherModel), {
        schema: {
            attributes: {
                payload: {
                    type: 'text',
                    required: true
                }
            }
        }
    })
        .addAlternativeName('standard-output');
    mainInstance.protocolManager.addProtocol(protocol);
}
exports.entryPoint = entryPoint;
