// enqueuer/js/code-generators/assertion-code-generator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class AssertionCodeGenerator {
    constructor(testsName, asserterInstanceName, assertionName) {
        this.testsName = testsName;
        this.asserterInstanceName = asserterInstanceName;
        this.assertionName = assertionName;
    }
    generate() {
        return `try {
            const evaluated = {name: assertion.name};
            (Object.keys(${this.assertionName}) || [])
                .filter(key => key !== 'name')
                .forEach(key =>       evaluated[key] = eval(${this.assertionName}[key])   );
            const testResult = ${this.asserterInstanceName}.assert(evaluated, ${this.assertionName});
            if (assertion.ignore !== undefined && assertion.ignore !== false) {
                testResult.ignored = true;
            }
            ${this.testsName}.push(testResult);
        } catch (err) {
            const msg = \`Error executing assertion: '\${err}'\`;
            Logger.error(msg);
            ${this.testsName}.push({
                name: assertion.name,
                description: msg,
                valid: false
            });
        }`;
    }
}
exports.AssertionCodeGenerator = AssertionCodeGenerator;
