// enqueuer/js/notifications/notification-emitter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const events_1 = require("events");
const logger_1 = require("../loggers/logger");
var Notifications;
(function (Notifications) {
    Notifications[Notifications["REQUISITION_RAN"] = 0] = "REQUISITION_RAN";
})(Notifications = exports.Notifications || (exports.Notifications = {}));
class NotificationEmitter {
    constructor() {
        this.eventEmitter = new events_1.EventEmitter();
    }
    static on(event, listener) {
        NotificationEmitter.notificationEmitter.eventEmitter.on(Notifications[event], listener);
        return NotificationEmitter.notificationEmitter;
    }
    static emit(event, ...args) {
        logger_1.Logger.trace(`Notification '${Notifications[event]}' emitted`);
        NotificationEmitter.notificationEmitter.eventEmitter.emit(Notifications[event], ...args);
        return NotificationEmitter.notificationEmitter;
    }
}
NotificationEmitter.notificationEmitter = new NotificationEmitter();
exports.NotificationEmitter = NotificationEmitter;
