// enqueuer/js/object-parser/yml-object-parser.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const yaml = __importStar(require("yamljs"));
const object_decycler_1 = require("./object-decycler");
class YmlObjectParser {
    parse(value) {
        return yaml.parse(value);
    }
    stringify(value, params = {}) {
        const parsedParams = this.parseParams(params);
        return yaml.stringify(new object_decycler_1.ObjectDecycler().decycle(value || {}), parsedParams.inline, parsedParams.space);
    }
    parseParams(params) {
        return Object.assign({}, {
            inline: 100,
            space: 2
        }, params);
    }
}
exports.YmlObjectParser = YmlObjectParser;
function entryPoint(mainInstance) {
    mainInstance.objectParserManager.addObjectParser(() => new YmlObjectParser(), 'yml', 'yaml');
}
exports.entryPoint = entryPoint;
