// enqueuer/js/reporters/subscription/subscription-final-reporter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class SubscriptionFinalReporter {
    constructor(summary) {
        this.messageReceivedTestName = `Message received`;
        this.subscriptionAvoidedTestName = `Subscription avoided`;
        this.noTimeOutTestName = `No time out`;
        this.subscribedTestName = `Subscribed`;
        this.hasMessage = false;
        this.subscribed = summary.subscribed;
        this.avoidable = summary.avoidable;
        this.hasMessage = summary.hasMessage;
        this.time = summary.time;
        this.subscribeError = summary.subscribeError;
        this.ignore = !!summary.ignore;
    }
    getReport() {
        if (this.ignore) {
            return [];
        }
        if (!this.subscribed) {
            return this.createNotSubscribedTests();
        }
        let tests = [];
        if (this.avoidable) {
            tests = tests.concat(this.createAvoidableTests());
        }
        else {
            tests = tests.concat(this.createMessageTests());
        }
        return tests.concat(this.addTimeoutTests());
    }
    addTimeoutTests() {
        if (this.time) {
            if (!!this.time.timeout && this.time.totalTime > this.time.timeout) {
                return this.createTimeoutTests();
            }
        }
        return [];
    }
    createNotSubscribedTests() {
        return [{
                valid: false,
                name: this.subscribedTestName,
                description: this.subscribeError || ''
            }];
    }
    createMessageTests() {
        if (this.hasMessage) {
            return [{
                    valid: true,
                    name: this.messageReceivedTestName,
                    description: `Subscription has received its message`
                }];
        }
        else {
            return [{
                    valid: false,
                    name: this.messageReceivedTestName,
                    description: `Subscription has not received its message`
                }];
        }
    }
    createTimeoutTests() {
        if (!this.avoidable && this.time) {
            return [{
                    valid: false,
                    name: this.noTimeOutTestName,
                    description: `Not avoidable subscription has timed out: ${this.time.totalTime} > ${this.time.timeout}`
                }];
        }
        return [];
    }
    createAvoidableTests() {
        if (this.hasMessage) {
            return [{
                    valid: false,
                    name: this.subscriptionAvoidedTestName,
                    description: `Avoidable subscription should not receive messages`
                }];
        }
        else {
            return [{
                    valid: true,
                    name: this.subscriptionAvoidedTestName,
                    description: `Avoidable subscription has not received any message`
                }];
        }
    }
}
exports.SubscriptionFinalReporter = SubscriptionFinalReporter;
