// enqueuer/js/pools/http-container.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const logger_1 = require("../loggers/logger");
const express_1 = __importDefault(require("express"));
const https_1 = __importDefault(require("https"));
const http_1 = __importDefault(require("http"));
const handler_listener_1 = require("../handlers/handler-listener");
class HttpContainer {
    constructor(port, credentials) {
        this.counter = 0;
        this.sockets = new Set();
        this.port = port;
        this.app = this.createApp();
        this.server = this.createServer(credentials);
        this.handleNewSocketConnections();
    }
    acquire() {
        return __awaiter(this, void 0, void 0, function* () {
            logger_1.Logger.debug(`Acquiring container ${this.port} (${this.counter})`);
            ++this.counter;
            if (this.counter == 1) {
                if (this.closePromise) {
                    yield this.closePromise;
                }
                this.listenPromise = new handler_listener_1.HandlerListener(this.server).listen(this.port);
            }
            yield this.listenPromise;
            logger_1.Logger.trace(`container acquired ${this.port} (${this.counter})`);
            return this.app;
        });
    }
    release() {
        return __awaiter(this, void 0, void 0, function* () {
            --this.counter;
            logger_1.Logger.trace(`Releasing container ${this.port} (${this.counter})`);
            if (this.counter == 0) {
                logger_1.Logger.trace(`Closing container ${this.port}`);
                this.closePromise = new Promise(((resolve, reject) => {
                    this.sockets.forEach((socket) => socket.destroy());
                    this.server.close((err) => {
                        if (err) {
                            reject(`Error closing server ${this.port}: ${err}`);
                        }
                        logger_1.Logger.trace(`Container ${this.port} is closed`);
                        resolve();
                    });
                }));
            }
            else {
                this.closePromise = Promise.resolve();
            }
            yield this.closePromise;
            return this.counter;
        });
    }
    handleNewSocketConnections() {
        this.server.on('connection', (socket) => {
            logger_1.Logger.trace(`Container ${this.port} got a new connection`);
            this.sockets.add(socket);
            socket.on('close', () => {
                this.sockets.delete(socket);
            });
        });
    }
    createServer(credentials) {
        if (credentials) {
            return https_1.default.createServer(credentials, this.app);
        }
        return http_1.default.createServer(this.app);
    }
    createApp() {
        const app = express_1.default();
        app.use((req, res, next) => {
            req.setEncoding('utf8');
            req.rawBody = '';
            req.on('data', (chunk) => {
                req.rawBody += chunk;
            });
            req.on('end', () => {
                next();
            });
            res.header('Access-Control-Allow-Origin', '*');
            res.header('Access-Control-Allow-Headers', 'Origin, X-Requested-With, Content-Type, Accept');
        });
        return app;
    }
}
exports.HttpContainer = HttpContainer;
